torch/__init__.py

```diff
@@ -778,3 +786,154 @@ def _register_device_module(device_type, module):
+        with autocast():
+            # torch.mm is on autocast's list of ops that should run in float16.
+            # Inputs are float32, but the op runs in float16 and produces float16 output.
+            # No manual casts are required.
+            e_float16 = torch.mm(a_float32, b_float32)
+            # Also handles mixed input types
+            f_float16 = torch.mm(d_float32, e_float16)
+        # After exiting autocast, calls f_float16.float() to use with d_float32
+        g_float32 = torch.mm(d_float32, f_float16.float())
+    Type mismatch errors *in* an autocast-enabled region are a bug; if this is what you observe,
+    please file an issue.
+    ``autocast(enabled=False)`` subregions can be nested in autocast-enabled regions.
+    Locally disabling autocast can be useful, for example, if you want to force a subregion
+    to run in a particular ``dtype``.  Disabling autocast gives you explicit control over
+    the execution type.  In the subregion, inputs from the surrounding region
+    should be cast to ``dtype`` before use::
+        # Creates some tensors in default dtype (here assumed to be float32)
+        a_float32 = torch.rand((8, 8), device="cuda")
+        b_float32 = torch.rand((8, 8), device="cuda")
+        c_float32 = torch.rand((8, 8), device="cuda")
+        d_float32 = torch.rand((8, 8), device="cuda")
+        with autocast():
+            e_float16 = torch.mm(a_float32, b_float32)
+            with autocast(enabled=False):
+                # Calls e_float16.float() to ensure float32 execution
+                # (necessary because e_float16 was created in an autocasted region)
+                f_float32 = torch.mm(c_float32, e_float16.float())
+            # No manual casts are required when re-entering the autocast-enabled region.
+            # torch.mm again runs in float16 and produces float16 output, regardless of input types.
+            g_float16 = torch.mm(d_float32, f_float32)
+    The autocast state is thread-local.  If you want it enabled in a new thread, the context manager or decorator
+    must be invoked in that thread.  This affects :class:`torch.nn.DataParallel` and
+    :class:`torch.nn.parallel.DistributedDataParallel` when used with more than one GPU per process
+    (see :ref:`Working with Multiple GPUs<amp-multigpu>`).
+    Args:
+        enabled(bool, optional, default=True):  Whether autocasting should be enabled in the region.
+    """
+    def __init__(self, enabled=True, fast_dtype=torch.float16, device='cuda'):
+        self.device = device
+        if device == 'cpu':
+            supported_dtype = [torch.bfloat16]
+            if fast_dtype not in supported_dtype :
+                warnings.warn("In CPU autocast, but the target dtype is not supported. Disable the autocast.")
+                warnings.warn("CPU Autocast only support dtype of torch.bfloat16 currently.")
+                enabled = False
+                fast_dtype = torch.bfloat16
```

Comment:
you should make sure that `device` is either `cuda` and `cpu`, and error out otherwise

Reply:
Instead of silently resetting user-supplied fast_dtype value you should error out if it's invalid

Reply:
Thank you for all of the suggestions, I will implement them and update the PR.